app/components/data-download.tsx

```diff
@@ -80,43 +75,81 @@ const DataDownloadInner = memo(
     observations,
     dimensions,
     measures,
-    fields,
+    extent,
+    fileFormat,
   }: {
     title: string;
     observations: Observation[];
     dimensions: DimensionMetaDataFragment[];
     measures: DimensionMetaDataFragment[];
-    fields: ChartFields;
+    extent: DataDownloadExtent;
+    fileFormat: DataDownloadFileFormat;
   }) => {
-    const forCsvData = useMemo(() => {
-      const columns = [...dimensions, ...measures];
-      return observations.map((obs) => {
-        return Object.keys(obs).reduce((acc, key) => {
-          const col = columns.find((d) => d.iri === key);
-
+    const data = useMemo(() => {
+      const columns = keyBy([...dimensions, ...measures], (d) => d.iri);
+      return observations.map((obs) =>
+        Object.keys(obs).reduce((acc, key) => {
+          const col = columns[key];
           return col
             ? {
                 ...acc,
                 ...{ [col.label]: obs[key] },
               }
             : acc;
-        }, {});
-      });
+        }, {})
+      );
     }, [dimensions, measures, observations]);
 
-    const csvData = csvFormat(forCsvData);
+    const onClick: () => void = useMemo(() => {
+      switch (fileFormat) {
+        case "csv":
+          const csvData = csvFormat(data);
+          const blob = new Blob([csvData], {
+            type: "text/plain;charset=utf-8",
+          });
+          return () => saveAs(blob, `${title}.csv`);
+        case "xlsx":
+          const workbook = XLSX.utils.book_new();
+          const worksheet = XLSX.utils.json_to_sheet(data);
+          XLSX.utils.book_append_sheet(workbook, worksheet, "data");
+          return () => XLSX.writeFile(workbook, `${title}.xlsx`);
+      }
+    }, [data, fileFormat, title]);
```

Comment:
I think a nice improvement would be to do it on the server, I think it would preferable as we would not ship a XLSX library to the client. The library seems a bit large : https://bundlephobia.com/package/xlsx@0.18.3.

Reply:
Done – I think there might be one obstacle moving forward, which is 4MB request size limit, I think it fair to say that there will be bigger datasets, especially when it comes to downloading full data. Not sure how we can proceed with that...